File: commodity-backend/routes/detail.py
from __future__ import annotations

from datetime import datetime, timedelta

from flask import Blueprint, request
from flask_jwt_extended import jwt_required

from models import Comment, Issue, Product, db
from models import issue_competitors_list, issue_trend_list
from utils.response import err, ok

# bp = Blueprint("detail", __name__, url_prefix="/detail")
bp = Blueprint("detail", __name__)


def to_iso_z(dt):
    if not dt:
        return ""
    s = dt.isoformat()
    if "Z" not in s and "+" not in s and "-" not in s[10:]:
        s = f"{s}Z"
    return s


@bp.route("/<issue_id>", methods=["GET"])
@jwt_required()
def issue_detail(issue_id: str):
    if not str(issue_id).isdigit():
        return err("无效的问题 ID", code=400, http_status=400)
    iid = int(issue_id)

    issue = db.session.get(Issue, iid)
    if not issue:
        return err("数据不存在", code=404, http_status=404)

    product = db.session.get(Product, int(issue.product_id)) if issue.product_id else None
    main_product_name = (product.name if product else "").strip() or "主商品"

    competitors = []
    for item in issue_competitors_list(issue):
        brand = str(item.get("brand") or "").strip() or "未知品牌"
        mentions = int(item.get("mentions", 0) or 0)
        sentiment = float(item.get("sentiment", 0) or 0)
        is_main = brand in ("本品", main_product_name)
        competitors.append(
            {
                "brand": brand,
                "displayName": f"主商品：{main_product_name}" if is_main else f"竞品：{brand}",
                "role": "main" if is_main else "competitor",
                "mentions": mentions,
                "sentiment": sentiment,
            }
        )

    data = {
        "id": str(issue.id),
        "name": issue.name,
        "mentions": issue.mentions,
        "sentiment": float(issue.sentiment_mean or 0),
        "severity": int(issue.severity),
        "productName": main_product_name,
        "trend": issue_trend_list(issue),
        "competitors": competitors,
    }
    return ok(data)


@bp.route("/<issue_id>/comments", methods=["GET"])
@jwt_required()
def issue_comments(issue_id: str):
    if not str(issue_id).isdigit():
        return err("无效的问题 ID", code=400, http_status=400)
    iid = int(issue_id)

    issue = db.session.get(Issue, iid)
    if not issue:
        return err("数据不存在", code=404, http_status=404)

    page = request.args.get("page", type=int) or 1
    page_size = request.args.get("pageSize", type=int) or 10
    page = max(1, page)
    page_size = max(1, min(100, page_size))

    q = Comment.query.filter(Comment.issue_id == iid).order_by(Comment.created_at.desc())
    total = q.count()
    items = q.offset((page - 1) * page_size).limit(page_size).all()

    lst = [
        {
            "id": str(c.id),
            "issueId": str(issue.id),
            "username": c.username or "匿名用户",
            "content": c.content,
            "sentiment": float(c.sentiment) if c.sentiment is not None else 0.5,
            "createdAt": to_iso_z(c.created_at),
        }
        for c in items
    ]
    return ok({"list": lst, "total": total})


@bp.route("/<issue_id>/batch-trend", methods=["GET"])
@jwt_required()
def issue_batch_trend(issue_id: str):
    """
    批次关联分析（轻量版）：
    - 使用评论里的 `raw_product_variant` 当作“批次/变体”维度
    - 返回近 days 天内，不同批次对该问题的提及次数趋势
    """
    if not str(issue_id).isdigit():
        return err("无效的问题 ID", code=400, http_status=400)
    iid = int(issue_id)

    issue = db.session.get(Issue, iid)
    if not issue:
        return err("数据不存在", code=404, http_status=404)

    days = request.args.get("days", type=int) or 14
    days = max(7, min(30, days))
    top_batches = request.args.get("topBatches", type=int) or 5
    top_batches = max(3, min(10, top_batches))

    end_dt = datetime.utcnow()
    start_dt = end_dt - timedelta(days=days - 1)

    q = Comment.query.filter(Comment.issue_id == iid, Comment.created_at.isnot(None), Comment.created_at >= start_dt)
    rows = q.all()

    # batch -> date -> mentions
    batch_date_counts: dict[str, dict[str, int]] = {}
    batch_totals: dict[str, int] = {}
    date_set: set[str] = set()

    for c in rows:
        batch_name = (c.raw_product_variant or "").strip() or "未标注"
        day_key = (c.created_at or end_dt).date().isoformat()
        date_set.add(day_key)

        batch_date_counts.setdefault(batch_name, {})
        batch_date_counts[batch_name][day_key] = batch_date_counts[batch_name].get(day_key, 0) + 1
        batch_totals[batch_name] = batch_totals.get(batch_name, 0) + 1

    # 只保留提及最多的 top batches
    batches_sorted = sorted(batch_totals.items(), key=lambda x: x[1], reverse=True)
    selected_batches = [b for b, _ in batches_sorted[:top_batches]]

    date_list = sorted(date_set)
    out_batches = []
    for b in selected_batches:
        counts_per_date = [batch_date_counts.get(b, {}).get(d, 0) for d in date_list]
        mentions_total = sum(counts_per_date)
        # 简单“新批次/旧批次”趋势：用前半段 vs 后半段对比
        mid = len(date_list) // 2
        first_sum = sum(counts_per_date[:mid])
        last_sum = sum(counts_per_date[mid:])
        change_pct = None
        if first_sum > 0:
            change_pct = round(((last_sum - first_sum) / first_sum) * 100, 2)
        out_batches.append(
            {
                "batchName": b,
                "mentions": mentions_total,
                "changePct": change_pct,
                "series": counts_per_date,
            }
        )

    return ok({"date": date_list, "batches": out_batches})

File: commodity-system-user/src/mock/detail.ts
import type { BatchTrendResponse, IssueDetail, PagedComments } from '@/types/detail'
import { delay } from '@/mock/utils'
import { loadDb } from '@/mock/db'

export async function mockGetDetailById(payload: { id: string }): Promise<IssueDetail> {
  await delay(360)
  const db = loadDb()
  const issue = db.issues.find((i) => i.id === payload.id)
  if (!issue) throw new Error('数据不存在')
  return {
    id: issue.id,
    name: issue.name,
    mentions: issue.mentions,
    sentiment: issue.sentiment,
    severity: issue.severity,
    productName: '示例主商品',
    trend: issue.trend,
    competitors: issue.competitors.map((c) => ({
      brand: c.brand,
      displayName: c.brand === '本品' ? '主商品：示例主商品' : `竞品：${c.brand}`,
      role: c.brand === '本品' ? 'main' : 'competitor',
      mentions: c.mentions,
      sentiment: c.sentiment,
    })),
  }
}

export async function mockGetComments(payload: {
  id: string
  page: number
  pageSize: number
}): Promise<PagedComments> {
  await delay(340)
  const db = loadDb()
  const issue = db.issues.find((i) => i.id === payload.id)
  if (!issue) throw new Error('数据不存在')
  const page = Math.max(1, Math.floor(payload.page))
  const pageSize = Math.max(1, Math.floor(payload.pageSize))
  const total = issue.comments.length
  const start = (page - 1) * pageSize
  const list = issue.comments.slice(start, start + pageSize)
  return { list, total }
}

export async function mockGetBatchTrend(payload: { id: string; days?: number; topBatches?: number }): Promise<BatchTrendResponse> {
  await delay(260)
  const db = loadDb()
  const issue = db.issues.find((i) => i.id === payload.id)
  if (!issue) throw new Error('数据不存在')

  const days = payload.days ?? 14
  const topBatches = payload.topBatches ?? 5

  // 用固定“批次A/B/C”模拟批次维度；系列形状按 issue.trend 做缩放，避免页面完全静态。
  const date = issue.trend.slice(-days).map((t) => t.date)
  const baseSeries = issue.trend.slice(-days).map((t) => t.value)

  const allBatches = [
    { batchName: '批次A', factor: 1.0 },
    { batchName: '批次B', factor: 0.72 },
    { batchName: '批次C', factor: 0.48 },
    { batchName: '批次D', factor: 0.34 },
    { batchName: '批次E', factor: 0.26 },
  ].slice(0, Math.max(1, topBatches))

  const batches = allBatches.map((b) => {
    const series = baseSeries.map((v) => Math.max(0, Math.round(v * b.factor * (0.85 + (v % 7) * 0.02))))
    const mid = Math.floor(series.length / 2)
    const first = series.slice(0, mid).reduce((a, c) => a + c, 0)
    const last = series.slice(mid).reduce((a, c) => a + c, 0)
    const changePct = first > 0 ? round2(((last - first) / first) * 100) : null
    const mentions = series.reduce((a, c) => a + c, 0)
    return { batchName: b.batchName, mentions, changePct, series }
  })

  return { date, batches }
}

function round2(n: number) {
  return Math.round(n * 100) / 100
}

